### PR TITLE
Hibernate and Spring JPA feature

### PR DESCRIPTION
**Pull request by JeovanHerrera**

### CODE REVIEW

File: src/main/java/com/jeovan/gymcrmsystem/daos/TraineeDao.java [hunk 1]
```diff
@@ -1,15 +1,22 @@
...
 import java.util.UUID;
 
 @Component
-public class TraineeDao extends SimpleDao<Trainee>{
+public interface TraineeDao extends JpaRepository<Trainee, UUID> {
 
-    public TraineeDao(Map<UUID, Trainee> storage) {
-        super(storage);
-    }
+    @Query("SELECT t FROM Trainee t JOIN t.user u WHERE u.username = ?1")
```
Comment (lfmartinezpantoja) [inline src/main/java/com/jeovan/gymcrmsystem/daos/TraineeDao.java]: Do you have any reason to use @Query when you can just use[ query name method](https://docs.spring.io/spring-data/jpa/reference/jpa/query-methods.html) and leave that spring create the query for you?

File: src/main/java/com/jeovan/gymcrmsystem/daos/TraineeDao.java [hunk 2]
```diff
@@ -1,15 +1,22 @@
...
+    Optional<Trainee> findByUsername(String username);
+    @Modifying
+    @Transactional
+    @Query("DELETE FROM Trainee t WHERE EXISTS (SELECT 1 FROM User u WHERE u.username = ?1 AND u = t.user)")
+    int deleteByUsername(String username);
```
Comment (lfmartinezpantoja) [inline src/main/java/com/jeovan/gymcrmsystem/daos/TraineeDao.java]: same comment here, as we are using spring jpa perhaps we can just use query name method to execute this delete operation and get rid @modifyng and @query

File: src/main/java/com/jeovan/gymcrmsystem/daos/TrainerDao.java [hunk 3]
```diff
@@ -1,15 +1,22 @@
...
+public interface TrainerDao extends JpaRepository<Trainer, UUID> {
+    @Query("SELECT t FROM Trainer t JOIN t.user u WHERE u.username = ?1")
+    Optional<Trainer> findByUsername(String username);
 
-    public TrainerDao(Map<UUID, Trainer> storage) {
-        super(storage);
-    }
+    @Modifying
+    @Transactional
+    @Query("DELETE FROM Trainer t WHERE EXISTS (SELECT 1 FROM User u WHERE u.username = ?1 AND u = t.user)")
```
Comment (lfmartinezpantoja) [inline src/main/java/com/jeovan/gymcrmsystem/daos/TrainerDao.java]: same comment here

File: src/main/java/com/jeovan/gymcrmsystem/services/TrainerServiceImpl.java [hunk 4]
```diff
@@ -3,49 +3,80 @@
...
+        return trainerDao.findByUsername(username).get();
+    }
+
+    @Override
+    public void deleteByUsername(String username) {
+        trainerDao.deleteByUsername(username);
+    }
+
+    @Secured("ADMIN")
+    public Trainer updatePassword(String username){
```
Comment (lfmartinezpantoja) [inline src/main/java/com/jeovan/gymcrmsystem/services/TrainerServiceImpl.java:67]: Just a quick question, why updatePassword is not part of the TrainerService interface?, if we want to use this in the future can be convenient have this in the abstraction.

File: src/main/java/com/jeovan/gymcrmsystem/services/TraineeServiceImpl.java [hunk 5]
```diff
@@ -2,49 +2,91 @@
...
+    @Secured("ADMIN")
+    public Trainee updatePassword(String username){
+        Trainee trainee = selectByUsername(username);
+        User user = trainee.getUser();
+        user.setPassword(passwordEncoder.encode(credentialGeneratorService.generatePassword()));
+        return traineeDao.save(trainee);
+    }
+
+    @Secured("ADMIN")
+    public Trainee toggleActiveStatus(String username){
```
Comment (lfmartinezpantoja) [inline src/main/java/com/jeovan/gymcrmsystem/services/TraineeServiceImpl.java:66]: I think toggleActiveStatus should be part of the interface so we can use it in the future regardless of the implementation

**Review by lfmartinezpantoja — CHANGES_REQUESTED** [review]
Hey Jeovainis good job, I left you some comments.

File: src/main/java/com/jeovan/gymcrmsystem/daos/TraineeDao.java (hunk 1 above)
Comment (JeovanHerrera) [reply src/main/java/com/jeovan/gymcrmsystem/daos/TraineeDao.java]: This was very silly, I was not aware I could use deleteByUserUsername and avoid having to create a Join, this was fixed in the next commit

File: src/main/java/com/jeovan/gymcrmsystem/daos/TraineeDao.java (hunk 2 above)
Comment (JeovanHerrera) [reply src/main/java/com/jeovan/gymcrmsystem/daos/TraineeDao.java]: Same as the other one, fixed!

File: src/main/java/com/jeovan/gymcrmsystem/services/TrainerServiceImpl.java (hunk 4 above)
Comment (JeovanHerrera) [reply src/main/java/com/jeovan/gymcrmsystem/services/TrainerServiceImpl.java:67]: I went backwards, first I implemented the method then added to the interface. Fixed in next commit

File: src/main/java/com/jeovan/gymcrmsystem/services/TraineeServiceImpl.java (hunk 5 above)
Comment (JeovanHerrera) [reply src/main/java/com/jeovan/gymcrmsystem/services/TraineeServiceImpl.java:66]: Added to interface

**Review by lfmartinezpantoja — APPROVED** [review]
Good job, as a comment apart, we can update the DataLoading and create some records in the database. In another hand update the unit test sounds pretty good, if need any help for service classes let me know.

**Pull request merged**